mlir_graphblas/ops.py

```diff
@@ -305,58 +305,131 @@ def call(cls, irbuilder, a, b, semiring, *, mask=None):
-            f"({input.type}) -> tensor<?x?xf64, #CSC64>"
+            f"{ret_val.assign} = call @cast_csx_to_csr({input}) : "
+            f"({input.type}) -> tensor<?x?xf64, {CSR64}>"
         )
 
 
-class CastCscToCsrOp(BaseOp):
+class CastCsxToCscOp(BaseOp):
     dialect = "util"
-    name = "cast_csc_to_csr"
+    name = "cast_csx_to_csc"
 
     @classmethod
     def call(cls, irbuilder, input):
         cls.ensure_mlirvar(input)
-        ret_val = irbuilder.new_var("tensor<?x?xf64, #CSR64>")
+        ret_val = irbuilder.new_var(f"tensor<?x?xf64, {CSC64}>")
         return ret_val, (
-            f"{ret_val.assign} = call @cast_csc_to_csr({input}) : "
-            f"({input.type}) -> tensor<?x?xf64, #CSR64>"
+            f"{ret_val.assign} = call @cast_csx_to_csc({input}) : "
+            f"({input.type}) -> tensor<?x?xf64, {CSC64}>"
+        )
+
+
+class PtrToTensorOp(BaseOp):
+    dialect = "util"
+    name = "ptr8_to_tensor"
+
+    @classmethod
+    def call(cls, irbuilder, input, return_type):
+        cls.ensure_mlirvar(input)
+        tensor_type = TensorType.parse(return_type, irbuilder.aliases)
+        encoding = tensor_type.encoding
+        if encoding.levels != ["dense", "compressed"]:
+            raise TypeError(
+                f"Return type must denote a CSR or CSC tensor (got {return_type})."
+            )
+        ret_val = irbuilder.new_var(f"tensor<?x?xf64, {CSX64}>")
+        ret_string = (
+            f"{ret_val.assign} = call @ptr8_to_tensor({input}) : "
+            f"(!llvm.ptr<i8>) -> tensor<?x?xf64, {CSX64}>"
+        )
+        if encoding.ordering is None:
+            pass
+        elif encoding.ordering == [0, 1]:
+            ret_val, cast_string = CastCsxToCsrOp.call(irbuilder, ret_val)
+            ret_string = ret_string + "\n" + cast_string
+        elif encoding.ordering == [1, 0]:
+            ret_val, cast_string = CastCsxToCscOp.call(irbuilder, ret_val)
+            ret_string = ret_string + "\n" + cast_string
+        else:
+            raise TypeError(
+                f"Return type must denote a CSR or CSC tensor (got {return_type})."
+            )
+        return ret_val, ret_string
+
+
+class TensorToPtrOp(BaseOp):
+    dialect = "util"
+    name = "tensor_to_ptr8"
+
+    @classmethod
+    def call(cls, irbuilder, input):
+        cls.ensure_mlirvar(input, TensorType)
+
+        encoding = input.type.encoding
+        if encoding.levels != ["dense", "compressed"]:
+            raise TypeError(
+                f"Input type must denote a CSR or CSC tensor (got {input.type})."
+            )
+
+        if encoding.ordering is None:
+            cast_string = ""
+        elif encoding.ordering == [0, 1]:
+            input, cast_string = CastCsrToCsxOp.call(irbuilder, input)
+            cast_string = "\n" + cast_string
+        elif encoding.ordering == [1, 0]:
+            input, cast_string = CastCscToCsxOp.call(irbuilder, input)
+            cast_string = "\n" + cast_string
```

Comment:
Shouldn't these be `cast_string = cast_string + "\n"`?

Reply:
Yes, that's correct. I updated it in https://github.com/metagraph-dev/mlir-graphblas/commit/31cdc28be5be6b4f6000e3411cbdd1fc30eee916.